In [1]:
from math import factorial

In [2]:
def combinations(n, k):
    """сочетания"""
    return int(factorial(n) / (factorial(n - k) * factorial(k)))


def permutations(n):
    """перестановки"""
    return int(factorial(n))


def arrangements(n, k):
    """размещения"""
    return int(factorial(n) / factorial(n - k))

### 1) Из колоды в 52 карты извлекаются случайным образом 4 карты. a) Найти вероятность того, что все карты – крести. б) Найти вероятность, что среди 4-х карт окажется хотя бы один туз.

#### а)

In [3]:
# кол-во сочетаний сколькими способами можно вытащить любые 4 карты из 52
n = combinations(52, 4)
n

270725

In [4]:
# кол-во сочетаний сколькими способами можно вытащить 4 карты из 13 крестей 
m = combinations(13, 4)
m

715

In [5]:
p = m / n
p * 100

0.26410564225690275

     Ответ: 0.26%

#### б)

In [6]:
def calc_combinations(a, b):
    a1 = combinations(*a)
    b1 = combinations(*b)
    return a1 * b1

In [7]:
# количество сочетаний 4 карт из 52, когда 1 карта - туз
# 1 туз из 4 И 3 оставшиеся карты из подмножества без тузов (52-4)
c1 = calc_combinations((4, 1), (48, 3))
c1

69184

In [8]:
# количество сочетаний 4 карт из 52, когда 2 карты - тузы
# 2 туза из 4 И 2 оставшиеся карты из подмножества без тузов (52-4)
c2 = calc_combinations((4, 2), (48, 2))
c2

6768

In [9]:
# количество сочетаний 4 карт из 52, когда 3 карты - тузы
# 3 туза из 4 И 1 оставшаяся карта из подмножества без тузов (52-4)
c3 = calc_combinations((4, 3), (48, 1))
c3

192

In [10]:
# количество сочетаний 4 карт из 52, когда 4 карты - тузы
# 4 туза из 4 И 0 оставшихся карт из подмножества без тузов (52-4)
c4 = calc_combinations((4, 4), (48, 0))
c4

1

In [11]:
m = c1 + c2 + c3 + c4

In [12]:
n = combinations(52, 4)
p = m / n
p * 100

28.126327454058547

     Ответ: 28.13%

### 2) На входной двери подъезда установлен кодовый замок, содержащий десять кнопок с цифрами от 0 до 9. Код содержит три цифры, которые нужно нажать одновременно. Какова вероятность того, что человек, не знающий код, откроет дверь с первой попытки?

In [13]:
# сколько всего комбинаций можно придумать из трех цифр от 0 до 9
a = combinations(10, 3)
a

120

In [14]:
p = 1 / a
p * 100

0.8333333333333334

    Ответ: 0.83%

### 3) В ящике имеется 15 деталей, из которых 9 окрашены. Рабочий случайным образом извлекает 3 детали. Какова вероятность того, что все извлеченные детали окрашены?

    Через вероятности каждой позиции

In [15]:
a1 = 9/15
a1

0.6

In [16]:
a2 = 8/14
a2

0.5714285714285714

In [17]:
a3 = 7/13
a3

0.5384615384615384

In [18]:
p = a1 * a2 * a3
p * 100

18.461538461538456

    Через сочетания

In [19]:
a = combinations(15, 3)
a

455

In [20]:
b = combinations(9, 3)
b

84

In [21]:
p = b / a
p * 100

18.461538461538463

    Ответ: 18.46%

### 4) В лотерее 100 билетов. Из них 2 выигрышных. Какова вероятность того, что 2 приобретенных билета окажутся выигрышными?

    Через вероятности каждой позиции

In [22]:
a1 = 2 / 100
a1

0.02

In [23]:
a2 = 1 / 99
a2

0.010101010101010102

In [24]:
p = a1 * a2
p * 100

0.020202020202020204

    Через сочетания

In [25]:
# общее кол-во вариантов, как моэно выбрать 2 билета из 100, без учета порядка (сочетания)
a = combinations(100, 2)
a

4950

In [26]:
p = 1 / a
p * 100

0.0202020202020202

    Ответ: 0.02 %